src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java

```diff
@@ -64,202 +122,3533 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
+          AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+          AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+          MFARequiredButNotConfiguredException {
+        LoginResult tutorLogin = loginAs(httpSession, TEST_TUTOR_EMAIL, TEST_TUTOR_PASSWORD);
+        HttpServletRequest getQuizAssignmentAttemptRequest = createRequestWithCookies(new Cookie[] {tutorLogin.cookie});
+        replay(getQuizAssignmentAttemptRequest);
+
+        Response getQuizAssignmentAttemptResponse =
+            quizFacade.getQuizAssignmentAttempt(getQuizAssignmentAttemptRequest, QUIZ_ASSIGNMENT_ID,
+                QUIZ_FACADE_TEST_STUDENT_2_ID_WITH_ASSIGNMENTS);
+
+        assertEquals(Response.Status.FORBIDDEN.getStatusCode(), getQuizAssignmentAttemptResponse.getStatus());
+
+        String actualErrorMessage = getQuizAssignmentAttemptResponse.readEntity(SegueErrorResponse.class).getErrorMessage();
+        assertEquals(INCORRECT_ROLE_ERROR_MESSAGE, actualErrorMessage);
+      }
+
+      @Test
+      public void asTeacher_whoIsNotGroupManager_isForbidden()
+          throws NoCredentialsAvailableException, NoUserException, SegueDatabaseException,
+          AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+          AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+          MFARequiredButNotConfiguredException {
+        LoginResult teacherLogin = loginAs(httpSession, DAVE_TEACHER_EMAIL, DAVE_TEACHER_PASSWORD);
+        HttpServletRequest getQuizAssignmentAttemptRequest =
+            createRequestWithCookies(new Cookie[] {teacherLogin.cookie});
+        replay(getQuizAssignmentAttemptRequest);
+
+        Response getQuizAssignmentAttemptResponse =
+            quizFacade.getQuizAssignmentAttempt(getQuizAssignmentAttemptRequest, QUIZ_ASSIGNMENT_ID,
+                QUIZ_FACADE_TEST_STUDENT_2_ID_WITH_ASSIGNMENTS);
+
+        assertEquals(Response.Status.FORBIDDEN.getStatusCode(), getQuizAssignmentAttemptResponse.getStatus());
+
+        String actualErrorMessage = getQuizAssignmentAttemptResponse.readEntity(SegueErrorResponse.class).getErrorMessage();
+        assertEquals("You can only view assignments to groups you own or manage.", actualErrorMessage);
+      }
+    }
+
+    @Nested
+    class CorrectGroupOwnerButInvalidTarget {
+      @Test
+      public void asTeacher_withTargetStudent_whoIsNotInGroup_isForbidden()
+          throws NoCredentialsAvailableException, NoUserException, SegueDatabaseException,
+          AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+          AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+          MFARequiredButNotConfiguredException {
+        LoginResult teacherLogin = loginAs(httpSession, TEST_TEACHER_EMAIL, TEST_TEACHER_PASSWORD);
```

Comment:
Another way you can reduce the size of this file and make it easier to read is by moving stuff that you keep repeating in multiple tests, like these three lines, into their own function. For instance, these three lines could be a function called `prepareTeacherRequest` (or something like that), which returns the `HttpServletRequest`, already replayed.